docs/tutorials/sessionkeys/sessionvalidationmodule.md

```diff
@@ -133,113 +333,181 @@ This function is essential for **validating user operations** against **session
+        uint256 permittedValueLimit = uint256(
+            uint128(bytes16(_sessionKeyData[44:60]))
         );
+        // rules list length is encoded as uint16 which is 2 bytes length
+        uint256 rulesListLength = uint256(
+            uint16(bytes2(_sessionKeyData[60:62]))
+        );
+
+        if (destinationContract != permittedDestinationContract) {
+            revert("ABISV Destination Forbidden");
+        }
+
+        if (bytes4(_funcCallData[0:4]) != permittedSelector) {
+            revert("ABISV Selector Forbidden");
+        }
+
+        if (callValue > permittedValueLimit) {
+            revert("ABISV Permitted Value Exceeded");
+        }
+
+        // avoided explicit check that (_sessionKeyData.length - 62) is the multiple of RULE_LENGTH
+        // also avoided calculating the rules list length from the rules list itself
+        // both to save on gas
+        // there is a test case that demonstrates that if the incorrect rules list length is provided
+        // the validation will fail
+        if (
+            !_checkRulesForPermission(
+                _funcCallData,
+                rulesListLength,
+                bytes(_sessionKeyData[62:]) //the rest of the _sessionKeyData is the rules list
+            )
+        ) {
+            revert("ABISV Arg Rule Violated");
+        }
 
-        require(recipient == recipientCalled, "ERC20SV Wrong Recipient");
-        require(amount <= maxAmount, "ERC20SV Max Amount Exceeded");
         return sessionKey;
     }
 ```
 
 **Operational Flow:**
 
-1. **Decode Session Key Data:** Extracts session key, token address, recipient address, and maximum token amount.
-2. **Validation Checks:** Ensures the destination contract and call value are as required.
-3. **Recipient and Amount Verification:** Compares recipient and transaction amount against session key data.
-4. **Return Session Key:** If all checks pass, returns the session key address.
+The method internally calls **_validateSessionParams** to perform the validation against the session key permissions.
+It passes the necessary parameters to **_validateSessionParams**  and returns the result.
+
+**_validateSessionParams** validates the parameters of the call against the permissions defined by the session key.
+
+**Execution flow in **_validateSessionParams****
+
+- Extracts essential details from _sessionKeyData, such as session key, permitted destination contract, selector, value limit, and rules list length.
+- Checks if the destinationContract matches the permitted destination contract, if the function selector in _funcCallData matches the permitted selector, and if the callValue exceeds the permitted value limit.
+- Calls **_checkRulesForPermission** to verify if the call data complies with the permission rules.
+
+**Execution flow in **_checkRulesForPermission****
+
+- Iterates through the rules list.
+- Parses each rule to determine the offset, condition, and value.
```

Comment:
Should we provide more information about `offset`, `condition` and `value`?

Reply:
### Yes. Let me write a bit more about them and I'll ask you to have a look at it again.